### PR TITLE
Import a smart contract into the NextJS UI project scaffold

### PR DESCRIPTION
**Pull request by ymekuria**
Addresses
#280 #291 

This PR adds the the template Add smart contract into the index file of the UI when NextJs is selected as the UI framework in the CLI. Smart contract integrations with Svelte and Nuxt will each follow in separate PRs.

### CODE REVIEW

File: src/lib/project.js [hunk 1]
```diff
@@ -137,6 +144,9 @@ async function project({ name, ui }) {
     `npm ci --silent > ${isWindows ? 'NUL' : '"/dev/null" 2>&1'}`
   );
 
+  // Build the template contract so it can be imported into the ui scaffold
```
Comment (jasongitmail) [inline src/lib/project.js:147]: seems reasonable to do this

File: src/lib/project.js [hunk 2]
```diff
@@ -323,10 +340,67 @@ async function scaffoldNext() {
         ],
       },
     ];
-  }
+  },
 };`
   );
+  newNextConfig = newNextConfig.replace(
```
Comment (jasongitmail) [inline src/lib/project.js:346]: Can we add a comment for why this is needed?

File: src/lib/project.js [hunk 3]
```diff
@@ -323,10 +340,67 @@ async function scaffoldNext() {
...
+    'reactStrictMode: true',
+    'reactStrictMode: false'
+  );
   fs.writeFileSync(path.join('ui', 'next.config.js'), newNextConfig);
+
+  const indexFileName = useTypescript === 'yes' ? 'index.tsx' : 'index.js';
```
Comment (jasongitmail) [inline src/lib/project.js:352]: should this be `.jsx`? Should be based off what NextJS tooling creates

File: src/lib/project.js [hunk 4]
```diff
@@ -323,10 +340,67 @@ async function scaffoldNext() {
...
+
+  fs.writeFileSync(
+    path.join('ui', 'pages', indexFileName),
+    newNextIndex,
+    'utf8'
+  );
+
+  // TODO: The command npm run build needs to be invoked in the smart contract directory root to generate
```
Comment (jasongitmail) [inline src/lib/project.js:360]: comment can be removed, but would be good to add a comment to the build step to note why it's needed (it's obvious, but also potentially useful to future devs reading this code)

File: src/lib/project.js [hunk 5]
```diff
@@ -323,10 +340,67 @@ async function scaffoldNext() {
...
+  // the build that is imported into the UI
+
+  newNextConfig = newNextConfig.replace(
```
Comment (jasongitmail) [inline src/lib/project.js:363]: Is this redundant with line 346? remove this probably

Comment (ymekuria) [inline src/lib/project.js:363]: remove redundancy

File: src/lib/project.js [hunk 6]
```diff
@@ -323,10 +340,67 @@ async function scaffoldNext() {
...
+    'reactStrictMode: true',
+    'reactStrictMode: false'
+  );
+  fs.writeFileSync(path.join('ui', 'next.config.js'), newNextConfig);
+
+  const tsconfig = `
```
Comment (jasongitmail) [inline src/lib/project.js:369]: Ok for now. Future, we can consider if we want to create the `contracts` dir first and read this file from it (less to maintain).

File: src/lib/project.js [hunk 7]
```diff
@@ -323,10 +340,67 @@ async function scaffoldNext() {
...
+    'reactStrictMode: true',
+    'reactStrictMode: false'
+  );
   fs.writeFileSync(path.join('ui', 'next.config.js'), newNextConfig);
+
+  const indexFileName = useTypescript === 'yes' ? 'index.tsx' : 'index.js';
+
+  fs.writeFileSync(
+    path.join('ui', 'pages', indexFileName),
+    newNextIndex,
```
Comment (jasongitmail) [inline src/lib/project.js:356]: might prefer `customNextIndex` as name

File: src/lib/ui/next/newIndex.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+} from 'snarkyjs';
+
+export default function Home() {
+  const [contract, setContract] = useState<Add | null>();
+  useEffect(() => {
+    (async () => {
+      await isReady;
+      const { Add } = await import('../../contracts/build/src/');
+
+      // This is the public key of the smart contract. Replace with the key of your deployed smart contract.
```
Comment (jasongitmail) [inline src/lib/ui/next/newIndex.js:21]: > // Update this to use the address (public key) for your zkApp account
> // To try it out, you can try this address for an example "Add" smart contract that we've deployed to Berkeley Testnet B62qqkb7hD1We6gEfrcqosKt9C398VLp1WXeTo1i9boPoqF7B1LxHg

File: src/lib/ui/next/newIndex.js [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+      const zkAppAdress = PublicKey.fromBase58(
+        'B62qqkb7hD1We6gEfrcqosKt9C398VLp1WXeTo1i9boPoqF7B1LxHg4'
+      );
+
+      const zkAppInstance = new Add(zkAppAdress);
+      setContract(zkAppInstance);
```
Comment (jasongitmail) [inline src/lib/ui/next/newIndex.js:27]: would remove line 27 until we establish conventions around this

**Review by jasongitmail — APPROVED** [review]
I've reviewed the code with Yoni without running.
Maybe @MartinMinkov can also review and run it to confirm please?

**Review by MartinMinkov — APPROVED** [review]
Tested the changes and it all checks out, thanks Yoni! :D Approving 

Just want to request that we should add a log statement in the console when the error is thrown related to SnarkyJS and empty keys. When booting Next up, it's a little unfriendly to get an unexplained error message so something directly in the browser will go a long way.

**Comment by ymekuria** [comment]
> Tested the changes and it all checks out, thanks Yoni! :D Approving
> 
> Just want to request that we should add a log statement in the console when the error is thrown related to SnarkyJS and empty keys. When booting Next up, it's a little unfriendly to get an unexplained error message so something directly in the browser will go a long way.

Good suggestions!

**Pull request merged**